src/librbd/io/ImageRequestWQ.cc

```diff
@@ -134,7 +136,7 @@ ssize_t ImageRequestWQ<I>::discard(uint64_t off, uint64_t len,
   if (r < 0) {
     return r;
   }
-  return len;
+  return std::min<uint64_t>(std::numeric_limits<int>::max(), len);
```

Comment:
Interesting, I see some inconsistency. In `librbd::Image::discard()` we return ` -EINVAL` if `len` is greater than max int, while in `rbd_discard()` we don't have this check.

Also, don't you think it is better to truncate the return value explicitly in API functions? For me the reason of truncating would be more clear then when reading the code. And it will make adding `rbd_discard2` (if we eventually want) simpler. And the return value of `io_work_queue->discard` may be used as a return value of `ssize_t rbd_writesame()`, which might give unexpected result.

Reply:
Good point -- I didn't realize that the C++ API already threw an error.